internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -804,21 +804,31 @@ func (r *Reconciler) reconcileReferencedTemplate(ctx context.Context, in reconci
 func createErrorWithoutObjectName(err error, obj client.Object) error {
 	var statusError *apierrors.StatusError
 	if errors.As(err, &statusError) {
+		var msg string
+		if statusError.Status().Message != "" {
+			if obj != nil {
+				msg = fmt.Sprintf("failed to create %s", obj.GetObjectKind().GroupVersionKind().GroupKind().String())
+			} else {
+				msg = "failed to create object"
+			}
+		}
 		if statusError.Status().Details != nil {
+			if msg != "" {
+				msg = fmt.Sprintf("%s: ", msg)
+			}
 			var causes []string
 			for _, cause := range statusError.Status().Details.Causes {
 				causes = append(causes, fmt.Sprintf("%s: %s: %s", cause.Type, cause.Field, cause.Message))
 			}
-			var msg string
 			if len(causes) > 0 {
-				msg = fmt.Sprintf("failed to create %s.%s: %s", statusError.Status().Details.Kind, statusError.Status().Details.Group, strings.Join(causes, " "))
+				msg = fmt.Sprintf("%sfailed to create %s.%s: %s", msg, statusError.Status().Details.Kind, statusError.Status().Details.Group, strings.Join(causes, " "))
```

Comment:
Should we just overwrite the entire msg if we have details?

`msg` seems to be a bit redundant when Message is != "".

Current we should get something like:

`failed to create object: failed to create DockerMachineTemplate.infrastructure.cluster.x-k8s.io: ...`
`failed to create DockerMachineTemplate: failed to create DockerMachineTemplate.infrastructure.cluster.x-k8s.io: ...`

Reply:
Can do - I just wasn't certain what other cases might include a message so I didn't want to assume it's redundant in all cases - it is in the current case though.

Reply:
To avoid misunderstandings. That's what I'm suggesting:
```
		var msg string
		if statusError.Status().Message != "" {
			if obj != nil {
				msg = fmt.Sprintf("failed to create %s", obj.GetObjectKind().GroupVersionKind().GroupKind().String())
			} else {
				msg = "failed to create object"
			}
		}
		if statusError.Status().Details != nil {
			var causes []string
			for _, cause := range statusError.Status().Details.Causes {
				causes = append(causes, fmt.Sprintf("%s: %s: %s", cause.Type, cause.Field, cause.Message))
			}
			if len(causes) > 0 {
				msg = fmt.Sprintf("failed to create %s.%s: %s", statusError.Status().Details.Kind, statusError.Status().Details.Group, strings.Join(causes, " "))
			} else {
				msg = fmt.Sprintf("failed to create %s.%s", statusError.Status().Details.Kind, statusError.Status().Details.Group)
			}
			// Replace the statusError message with the constructed message.
		}
		statusError.ErrStatus.Message = msg
		return statusError
```

The only thing we would overwrite is the part of `msg` that we wrote ourselves in the first `if`. Probably should add a short comment that we do this intentional.

We could also flip the order and do something like this to make it very obvious
```go
		var msg string
		if statusError.Status().Details != nil {
			var causes []string
			for _, cause := range statusError.Status().Details.Causes {
				causes = append(causes, fmt.Sprintf("%s: %s: %s", cause.Type, cause.Field, cause.Message))
			}
			if len(causes) > 0 {
				msg = fmt.Sprintf("failed to create %s.%s: %s", statusError.Status().Details.Kind, statusError.Status().Details.Group, strings.Join(causes, " "))
			} else {
				msg = fmt.Sprintf("failed to create %s.%s", statusError.Status().Details.Kind, statusError.Status().Details.Group)
			}
			statusError.ErrStatus.Message = msg
			return statusError
		}
		
		if statusError.Status().Message != "" {
			if obj != nil {
				msg = fmt.Sprintf("failed to create %s", obj.GetObjectKind().GroupVersionKind().GroupKind().String())
			} else {
				msg = "failed to create object"
			}
		}
		statusError.ErrStatus.Message = msg
		return statusError
```

Reply:
Got it - should we instead just never return the err in error.Wrap? so if we don't have details we can always return the failed to create object message?

Reply:
Problem is that it would make it very hard to root-cause issues if I'm not missing anything. Essentially you would only now that creating failed, but you have no idea what's wrong and have to reproduce it somehow manually.

Reply:
Agreed - the current version is what you've suggested above. 

I think we probably do have gaps in how much information our conditions put out, but users should be able to glean enough from causes (the most common case) to understand their spec is wrong at this point.

Reply:
Yup. Just create doesn't work somehow and yes probably it's somewhere in the Cluster or ClusterClass isn't really good enough to debug it easily (although there's a small chance it even is something else).

Reply:
Sorry misread your comment.

With causes it's fine. I was refering to a case where we would only surface "failed to create <kind>"

Reply:
The final return for this func - if it's not a status error with the current implementation - is errors.Wrap() with the original error.

Reply:
Not sure if we lose information if we do it instead of the new if block. But if we keep the new if and only want to change the line, sounds fine to me.

btw, looks like we're not handling obj == nil there? (even though we do in if)

Reply:
I've updated this case to handle similarly to the above i.e. never return the object name. If it's a nil object we return a generic message, otherwise we return a completely new error with just the group kind.

Reply:
Sounds good. I'll resolve